### PR TITLE
Cache successful shard deletion checks

### PR DESCRIPTION
**Pull request by ywelsch**
Each node checks on every cluster state update if there are shards that it can possibly delete from its disk. It decides this by doing a file-system lookup for each shard id that is fully allocated in the cluster. With lots of shards, this amounts to lots of `Files.exists()` checks, considerably slowing down cluster state updates. This commit adds a caching layer so that the `Files.exists()` checks can be skipped if not needed.

I've done some simple benchmarking to validate the approach using the following test method:

```
public void testSimple() {
    String node1 = internalCluster().startNode();
    String node2 = internalCluster().startNode();
    ensureStableCluster(2);

    ClusterState state = internalCluster().getInstance(ClusterService.class, node1).state();
    ClusterStateChanges cluster = new ClusterStateChanges();
    for (int i = 0; i < 1000; i++) {
        String name = "index_" + i;
        Settings.Builder settingsBuilder = Settings.builder()
            .put(SETTING_NUMBER_OF_SHARDS, 1)
            .put(SETTING_NUMBER_OF_REPLICAS, 0)
            .put("index.routing.allocation.exclude._name", node1);
        CreateIndexRequest request = new CreateIndexRequest(name, settingsBuilder.build()).waitForActiveShards(ActiveShardCount.NONE);

        state = cluster.createIndex(state, request);
    }

    state = cluster.applyStartedShards(state, state.routingTable().shardsWithState(ShardRoutingState.INITIALIZING));

    ClusterState state1 = cluster.closeIndices(state, new CloseIndexRequest("index_0"));
    ClusterState state2 = cluster.openIndices(state, new OpenIndexRequest("index_0"));

    ClusterChangedEvent changedEvent = new ClusterChangedEvent("simulated change 1", state1, state2);

    IndicesStore indicesStore = internalCluster().getInstance(IndicesStore.class, node1);

    logger.info("first iteration");
    indicesStore.clusterChanged(changedEvent);
    logger.info("second iteration");
    indicesStore.clusterChanged(changedEvent);
    logger.info("finished");
}
```

Output on my Laptop (with SSD):

WITHOUT PATCH
```
[2016-11-10T02:30:15,456][INFO ][o.e.i.s.IndicesStoreIntegrationIT] first iteration
[2016-11-10T02:30:15,834][INFO ][o.e.i.s.IndicesStoreIntegrationIT] second iteration
[2016-11-10T02:30:16,015][INFO ][o.e.i.s.IndicesStoreIntegrationIT] finished
```

WITH PATCH APPLIED
```
[2016-11-10T13:29:07,053][INFO ][o.e.i.s.IndicesStoreIntegrationIT] first iteration
[2016-11-10T13:29:07,337][INFO ][o.e.i.s.IndicesStoreIntegrationIT] second iteration
[2016-11-10T13:29:07,340][INFO ][o.e.i.s.IndicesStoreIntegrationIT] finished
```

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 1]
```diff
@@ -128,15 +154,24 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+                    IndicesService.ShardDeletionCheckResult shardDeletionCheckResult = indicesService.canDeleteShardContent(shardId, indexSettings);
+                    switch (shardDeletionCheckResult) {
+                        case CAN_DELETE:
+                            deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                            break;
+                        case NO_FOLDER_FOUND:
+                        case SHARED_FILE_SYSTEM:
+                            deletionSkipCache.add(shardId);
+                            break;
+                        default:
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: Can we add a `case` for `NO_NODE_LOCK` and `STILL_ALLOCATED` that does a `break;` and then have the `default:` case throw an exception, just so in the future if we add another state it doesn't fall through?

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 2]
```diff
@@ -128,15 +154,24 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+                            // do nothing;
                     }
                 }
             }
         }
     }
 
-    boolean shardCanBeDeleted(ClusterState state, IndexShardRoutingTable indexShardRoutingTable) {
+    boolean shardCanBeDeleted(String localNodeId, IndexShardRoutingTable indexShardRoutingTable) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: Yay I believe this can be `static` now?

**Review by dakrone — APPROVED** [review]
Left a couple of comments, LGTM

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 2 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: correct :-)

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: good idea

File: core/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 3]
```diff
@@ -785,39 +785,50 @@ public IndexMetaData verifyIndexIsDeleted(final Index index, final ClusterState
...
                     // lets see if it's path is available (return false if the shared doesn't exist)
                     // we don't need to delete anything that is not there
-                    return FileSystemUtils.exists(nodeEnv.availableShardPaths(shardId));
+                    return FileSystemUtils.exists(nodeEnv.availableShardPaths(shardId)) ?
+                        ShardDeletionCheckResult.CAN_DELETE :
+                        ShardDeletionCheckResult.NO_FOLDER_FOUND;
                 }
-            }
+            } else {
+               return ShardDeletionCheckResult.NO_NODE_LOCK;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/IndicesService.java]: ending up here means the node doesn't have local storage. Can we rename this enum? (see DiscoveryNode#nodeRequiresLocalStorage and NodeEnvironment#NodeEnvironment)

File: core/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 4]
```diff
@@ -703,7 +703,7 @@ public void deleteShardStore(String reason, ShardId shardId, ClusterState cluste
         final IndexMetaData metaData = clusterState.getMetaData().indices().get(shardId.getIndexName());
 
         final IndexSettings indexSettings = buildIndexSettings(metaData);
-        if (canDeleteShardContent(shardId, indexSettings) == false) {
+        if (canDeleteShardContent(shardId, indexSettings) != ShardDeletionCheckResult.CAN_DELETE) {
             throw new IllegalStateException("Can't delete shard " + shardId);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/IndicesService.java]: can we add the result of canDeleteShardContent to the message?

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 5]
```diff
@@ -115,11 +123,29 @@ public void clusterChanged(ClusterChangedEvent event) {
             return;
         }
 
-        for (IndexRoutingTable indexRoutingTable : event.state().routingTable()) {
+        RoutingTable routingTable = event.state().routingTable();
+
+        // remove entries from cache that don't exist in the routing table anymore
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: can we add a note that these are either closed indices (shouldn't be deleted) or deleted indices that are dealt with somewhere else (link to where)?

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 6]
```diff
@@ -115,11 +123,29 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+        for (Iterator<ShardId> it = deletionSkipCache.iterator(); it.hasNext(); ) {
+            ShardId shardId = it.next();
+            if (routingTable.hasIndex(shardId.getIndex()) == false) {
+                it.remove();
+            }
+        }
+        // remove entries from cache which are allocated to this node
+        final String localNodeId = event.state().nodes().getLocalNodeId();
+        RoutingNode localRoutingNode = event.state().getRoutingNodes().node(localNodeId);
+        if (localRoutingNode != null) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: :(

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: I'm wondering if we should make IndicesStore be called from IndicesClusterStateService rather than making it a generic listener? As follow up of course.

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 7]
```diff
@@ -146,27 +185,12 @@ boolean shardCanBeDeleted(ClusterState state, IndexShardRoutingTable indexShardR
...
-                if (node == null) {
-                    return false;
-                }
-            }
-
-            // check if shard is active on the current node or is getting relocated to the our node
-            String localNodeId = state.getNodes().getLocalNode().getId();
-            if (localNodeId.equals(shardRouting.currentNodeId()) || localNodeId.equals(shardRouting.relocatingNodeId())) {
+            // check if shard is active on the current node
+            if (localNodeId.equals(shardRouting.currentNodeId())) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: why did you remove the relocation logic?

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 8]
```diff
@@ -128,15 +154,28 @@ public void clusterChanged(ClusterChangedEvent event) {
                     } else {
                         indexSettings = indexService.getIndexSettings();
                     }
-                    if (indicesService.canDeleteShardContent(shardId, indexSettings)) {
-                        deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                    IndicesService.ShardDeletionCheckResult shardDeletionCheckResult = indicesService.canDeleteShardContent(shardId, indexSettings);
+                    switch (shardDeletionCheckResult) {
+                        case CAN_DELETE:
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: I wish we somehow worked "Exists locally but can be deleted" into the name. How about `FOLDER_FOUND_CAN_DELETE` ? (it's long, I know :))

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 9]
```diff
@@ -128,15 +154,28 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+                            deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                            break;
+                        case NO_FOLDER_FOUND:
+                        case SHARED_FILE_SYSTEM:
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: I wonder if we should mix shared file systems into this. It makes the logic more complex and it doesn't really save us much because canDeleteShardContent is super fast in this case (it's a setting lookup). If remove it, we can call the cache map something like `folderNotFoundCache` which imo will be easier to follow (other names welcome :))

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 10]
```diff
@@ -146,27 +185,12 @@ boolean shardCanBeDeleted(ClusterState state, IndexShardRoutingTable indexShardR
...
             // be conservative here, check on started, not even active
-            if (!shardRouting.started()) {
+            if (shardRouting.started() == false) {
                 return false;
             }
 
-            // if the allocated or relocation node id doesn't exists in the cluster state  it may be a stale node,
-            // make sure we don't do anything with this until the routing table has properly been rerouted to reflect
-            // the fact that the node does not exists
-            DiscoveryNode node = state.nodes().get(shardRouting.currentNodeId());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: it's a happy day we can remove this and because our nodes and routing table is consistent. We should probably change this in deleteShardIfExistElseWhere:

```
            // Node can't be null, because otherwise shardCanBeDeleted() would have returned false
            DiscoveryNode currentNode = state.nodes().get(shardRouting.currentNodeId());
            assert currentNode != null;
```

Also - do we have assertions for this ClusterState builder, so we're sure it never happens?

**Review by bleskes — CHANGES_REQUESTED** [review]
Love the new enum result of canDeleteShardContent. Left some comments.

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 7 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: The shard is in state STARTED (i.e. not RELOCATING), so it cannot have a relocatingNodeId.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: active() != started(). Doh :)

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 9 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: It's not super fast. The IndexSettings object has to be created first just to do this check. I've seen cases where constructing this object again and again is non-negligible.

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 10 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: I don't think we have any assertion for this. The tricky thing is finding a good place for it (might be a bit too expensive to check it whenever a new cluster state is built).

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 8 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: ok

File: core/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 3 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/IndicesService.java]: this should never happen actually as IndicesStore is only run on data nodes and data nodes require local storage. I've added an `assert false`.

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/IndicesService.java]: I've added the `assert false` to IndicesStore, not here.

File: core/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 4 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/IndicesService.java]: done

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 5 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: done

**Comment by ywelsch** [comment]
@bleskes I've updated the PR. Please have another look.

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java [hunk 11]
```diff
@@ -128,15 +156,29 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+                    IndicesService.ShardDeletionCheckResult shardDeletionCheckResult = indicesService.canDeleteShardContent(shardId, indexSettings);
+                    switch (shardDeletionCheckResult) {
+                        case FOLDER_FOUND_CAN_DELETE:
+                            deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                            break;
+                        case NO_FOLDER_FOUND:
+                            folderNotFoundCache.add(shardId);
+                            break;
+                        case NO_LOCAL_STORAGE:
+                            assert false : "shard deletion only runs on data nodes which always have local storage";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: ++ to doing this here.

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 10 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: can you clarify what you mean by too expensive? I was talking about an assertion. Gut feeling says to put it on the ClusterState constructor ...

**Review by bleskes — APPROVED** [review]
LGTM. Thx @ywelsch

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 10 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: yes, I was as well thinking about an assertion. Putting it on the ClusterState constructor however means that this check will be done a lot of times. We often just make simple changes on cluster states like increasing version or setting local node which could slow down our tests. We have similar checks currently in `RoutingTable.validate` but this is only called after changes made by AllocationService. If we could ensure that DiscoveryNodes and RoutingTable/RoutingNodes are only changed in AllocationService (something which I think is feasible and on my stack of things to investigate), a check in that method would be enough. I'll investigate where to put the assertion (it's complementary to this PR).

**Comment by ywelsch** [comment]
@bleskes @dakrone I've pushed 647c84b which removes a bit more dead code.

File: core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java (hunk 10 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java]: >  I'll investigate where to put the assertion (it's complementary to this PR).

++

re the speed - I run IndicesClusterStateServiceRandomUpdatesTests with this patch in it (i.e., 100 times and the test it self iterates too). The performance difference is negligible (46s vs 44s - which is just random noise). I don't think we need to worry much about performance here.

```
diff --git a/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java b/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java
index 074b4a5..2016dad 100644
--- a/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java
+++ b/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java
@@ -19,6 +19,7 @@

 package org.elasticsearch.indices.cluster;

+import com.carrotsearch.randomizedtesting.annotations.Repeat;
 import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.admin.cluster.reroute.ClusterRerouteRequest;
@@ -71,6 +72,7 @@ import static org.elasticsearch.cluster.routing.ShardRoutingState.INITIALIZING;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;

+@Repeat(iterations = 100)
 public class IndicesClusterStateServiceRandomUpdatesTests extends AbstractIndicesClusterStateServiceTestCase {

     private final ClusterStateChanges cluster = new ClusterStateChanges();
@@ -341,6 +343,15 @@ public class IndicesClusterStateServiceRandomUpdatesTests extends AbstractIndice

         // TODO: go masterless?

+        for (ShardRouting shard: state.routingTable().allShards()) {
+            if (shard.assignedToNode()) {
+                assert state.nodes().get(shard.currentNodeId()) != null;
+            }
+            if (shard.relocating()) {
+                assert state.nodes().get(shard.relocatingNodeId()) != null;
+            }
+        }
+
         return state;
     }
```

**Comment by bleskes** [comment]
@ywelsch last commit looks good too.

**Comment by ywelsch** [comment]
Thanks for the reviews @dakrone @bleskes
